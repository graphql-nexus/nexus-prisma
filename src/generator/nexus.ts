import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
