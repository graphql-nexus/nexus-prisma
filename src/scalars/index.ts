export * from './DateTime'
export * from './Json'
