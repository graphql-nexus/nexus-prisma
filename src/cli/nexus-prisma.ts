#!/usr/bin/env node

/** This script will be run by the prisma generator system. */

process.env.DEBUG_COLORS = 'true'
process.env.DEBUG_HIDE_DATE = 'true'
import { generatorHandler } from '@prisma/generator-helper'
import * as Path from 'path'
import { generateRuntimeAndEmit } from '../generator'
import { loadUserGentimeSettings } from '../generator/gentime/settingsLoader'
import { Gentime } from '../generator/gentime/settingsSingleton'
import { externalToInternalDmmf } from '../helpers/prismaExternalToInternalDMMF'

// todo by default error in ci and warn in local
// enforceValidPeerDependencies({
//   packageJson: require('../../package.json'),
// })

generatorHandler({
  onManifest() {
    return {
      defaultOutput: Path.join(__dirname, '../runtime'),
      prettyName: 'Nexus Prisma',
    }
  },
  // async required by interface
  // eslint-disable-next-line
  async onGenerate({ dmmf, otherGenerators }) {
    const prismaClientGenerator = otherGenerators.find((g) => g.provider.value === 'prisma-client-js')

    if (!Gentime.settings.data.prismaClientLocation) {
      Gentime.changeSettings({
        prismaClientLocation: prismaClientGenerator?.output?.value || '@prisma/client',
      })
    }

    const internalDMMF = externalToInternalDmmf(dmmf)
    loadUserGentimeSettings()
    generateRuntimeAndEmit(internalDMMF, Gentime.settings)
    process.stdout.write(
      `You can now start using Nexus Prisma in your code. Reference: https://pris.ly/d/nexus-prisma\n`
    )
  },
})
