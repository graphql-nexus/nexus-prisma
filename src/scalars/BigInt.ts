import { BigIntResolver } from 'graphql-scalars'
import { asNexusMethod } from 'nexus'

/**
 * A Nexus scalar type definition for the `BigInt` scalar type
 *
 * Contributes a scalar to your GraphQL schema called `BigInt`.
 *
 * Contributes a `t` `[1]` helper method called `bigInt`
 *
 * `[1]` A `t` helper method refers to a method on the argument given to a `definition` method. Helper methods
 * here typically help you quickly create new fields.
 *
 * @example
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import { BigInt } from 'nexus-prisma/scalars'
 *
 *   SomeObject = objectType({
 *     name: 'SomeObject',
 *     definition(t) {
 *       t.bigInt('someBigIntField')
 *     },
 *   })
 *
 *   makeSchema({
 *     types: [BigInt, SomeObject],
 *   })
 *
 */
export const BigInt = asNexusMethod(BigIntResolver, 'bigInt')
