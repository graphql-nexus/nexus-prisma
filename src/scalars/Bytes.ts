import { GraphQLScalarType } from 'graphql'
import { ByteResolver } from 'graphql-scalars'
import { asNexusMethod } from 'nexus'

/**
 * A Nexus scalar type definition for the `Bytes` scalar type represents byte value as specified by [NodeJS
 * Buffer type](https://nodejs.org/api/buffer.html)
 *
 * Contributes a scalar to your GraphQL schema called `Bytes`.
 *
 * Contributes a `t` `[1]` helper method called `bytes`
 *
 * `[1]` A `t` helper method refers to a method on the argument given to a `definition` method. Helper methods
 * here typically help you quickly create new fields.
 *
 * @example
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import { Bytes } from 'nexus-prisma/scalars'
 *
 *   SomeObject = objectType({
 *     name: 'SomeObject',
 *     definition(t) {
 *       t.bytes('someBytesField')
 *     },
 *   })
 *
 *   makeSchema({
 *     types: [Bytes, SomeObject],
 *   })
 *
 */
export const Bytes = asNexusMethod(
  new GraphQLScalarType({
    ...ByteResolver,
    // Override the default 'Byte' name with one that matches what Nexus Prisma expects.
    name: 'Bytes',
  }),
  'bytes',
)
