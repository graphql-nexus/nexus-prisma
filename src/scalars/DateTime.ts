import { GraphQLScalarType } from 'graphql'
import { DateTimeResolver } from 'graphql-scalars'
import { asNexusMethod } from 'nexus'

export const dateTimeScalar = asNexusMethod(new GraphQLScalarType(DateTimeResolver), 'dateTime')
