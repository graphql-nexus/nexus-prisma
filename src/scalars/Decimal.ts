import { asNexusMethod } from 'nexus'
import * as decimal from 'decimal.js'

import { GraphQLScalarType, Kind } from 'graphql'

/**
 * A Nexus scalar type definition for arbitrary-precision Decimal type
 *
 * Contributes a scalar to your GraphQL schema called `Decimal`.
 *
 * Contributes a `t` `[1]` helper method called `decimal`
 *
 * `[1]` A `t` helper method refers to a method on the argument given to a `definition` method. Helper methods
 * here typically help you quickly create new fields.
 *
 * @example
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import { Decimal } from 'nexus-prisma/scalars'
 *
 *   SomeObject = objectType({
 *     name: 'SomeObject',
 *     definition(t) {
 *       t.decimal('someDecimalField')
 *     },
 *   })
 *
 *   makeSchema({
 *     types: [Decimal, SomeObject],
 *   })
 *
 */
export const Decimal = asNexusMethod(
  /**
   * Copied from prisma-graphql-type-decimal.
   *
   * @see https://github.com/unlight/prisma-graphql-type-decimal
   */
  new GraphQLScalarType({
    name: 'Decimal',
    description: 'An arbitrary-precision Decimal type',
    /**
     * Value sent to the client
     */
    serialize(value: decimal.Decimal) {
      return value.toString()
    },
    /**
     * Value from the client
     */
    parseValue(value: decimal.Decimal.Value) {
      return new decimal.Decimal(value)
    },
    parseLiteral(ast) {
      if (ast.kind === Kind.INT || ast.kind === Kind.FLOAT || ast.kind === Kind.STRING) {
        return new decimal.Decimal(ast.value)
      }
      return null
    },
  }),
  'decimal'
)
