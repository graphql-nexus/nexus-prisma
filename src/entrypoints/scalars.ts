import { BigInt } from '../scalars/BigInt'
import { Bytes } from '../scalars/Bytes'
import { DateTime } from '../scalars/DateTime'
import { Decimal } from '../scalars/Decimal'
import { Json } from '../scalars/Json'

/**
 * Predefined Nexus scalar type definitions to satisfy all custom scalars needed in GraphQL to map to the
 * native scalars in Prisma. The mapping is as follows:
 *
 * | Prisma     | GraphQL    | Nexus `t` Helper | GraphQL Scalar Implementation                                         | Additional Info                                                                                              |
 * | ---------- | ---------- | ---------------- | --------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
 * | `Json`     | `Json`     | `json`           | [JsonObject](https://www.graphql-scalars.dev/docs/scalars/jsonobject) |                                                                                                              |
 * | `DateTime` | `DateTime` | `dateTime`       | [DateTime](https://www.graphql-scalars.dev/docs/scalars/datetime)     |                                                                                                              |
 * | `BigInt`   | `BigInt`   | `bigInt`         | [BigInt](https://www.graphql-scalars.dev/docs/scalars/big-int)        | [JavaScript BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) |
 * | `Bytes`    | `Bytes`    | `bytes`          | [Byte](https://www.graphql-scalars.dev/docs/scalars/byte/)            | [Node.js Buffer](https://nodejs.org/api/buffer.html#buffer_buffer)                                           |
 * | `Decimal` | `Decimal` | `decimal` | (internal) | Uses [Decimal.js](https://github.com/MikeMcl/decimal.js) |.
 *
 * @example
 *
 *   // Use this default export
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import NexusPrismaScalars from 'nexus-prisma/scalars'
 *
 *   makeSchema({
 *     types: [NexusPrismaScalars],
 *   })
 *
 * @example
 *
 *   // Use ESM namespace import if you prefer
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import * as NexusPrismaScalars from 'nexus-prisma/scalars'
 *
 *   makeSchema({
 *     types: [NexusPrismaScalars],
 *   })
 *
 * @example
 *
 *   // Use only select predefined custom scalars
 *
 *   import { makeSchema, objectType } from 'nexus'
 *   import { Json } from 'nexus-prisma/scalars'
 *
 *   makeSchema({
 *     types: [Json],
 *   })
 *
 * @example
 *
 *   // Use your own custom scalars instead of these.
 *
 *   import { GraphQLScalarType } from 'graphql'
 *   import { JSONObjectResolver, DateTimeResolver } from 'graphql-scalars'
 *   import { asNexusMethod, makeSchema } from 'nexus'
 *
 *   const jsonScalar = new GraphQLScalarType({
 *     ...JSONObjectResolver,
 *     // Override the default 'JsonObject' name with one that matches what Nexus Prisma expects.
 *     name: 'Json',
 *   })
 *
 *   const dateTimeScalar = new GraphQLScalarType(DateTimeResolver)
 *
 *   makeSchema({
 *     types: [asNexusMethod(jsonScalar, 'json'), asNexusMethod(dateTimeScalar, 'dateTime')],
 *   })
 *
 * @remarks Some Of the Prisma scalars do not have a natural standard representation in GraphQL. For
 *          these cases Nexus Prisma generates code that references type names matching those scalar
 *          names in Prisma. Then, you are expected to define those custom scalar types in your GraphQL
 *          API. For convenience you can use these ones.
 */
const NexusPrismaScalars = {
  BigInt,
  Bytes,
  DateTime,
  Decimal,
  Json,
}

export default NexusPrismaScalars

export { BigInt, Bytes, DateTime, Decimal, Json }
