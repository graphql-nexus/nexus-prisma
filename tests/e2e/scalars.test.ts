import * as Execa from 'execa'
import stripAnsi from 'strip-ansi'
import { assertBuildPresent, createPrismaSchema, setupTestProject, TestProject } from '../__helpers__'

function setupTestProjectCase({
  prismaSchema,
  main,
  testProject,
}: {
  testProject: TestProject
  prismaSchema: string
  main: string
}) {
  testProject.tmpdir.write('main.ts', main)
  testProject.tmpdir.write('prisma/schema.prisma', prismaSchema)
}

interface ProjectResult {
  runFirstBuild: Execa.ExecaSyncReturnValue<string>
  runReflectPrisma: Execa.ExecaSyncReturnValue<string>
  runReflectNexus: Execa.ExecaSyncReturnValue<string>
  runSecondBuildStdout: Execa.ExecaSyncReturnValue<string>
  graphqlSchema?: string
  typegen?: string
}

function runTestProject(testProject: TestProject): ProjectResult {
  const commandConfig: Execa.SyncOptions = {
    reject: false,
    cwd: testProject.tmpdir.cwd(),
  }
  const runFirstBuild = Execa.commandSync(`npm run build`, commandConfig)
  const runReflectPrisma = Execa.commandSync(`npm run reflect:prisma`, commandConfig)
  const runReflectNexus = Execa.commandSync(`npm run reflect:nexus`, commandConfig)
  const runSecondBuildStdout = Execa.commandSync(`npm run build`, commandConfig)
  const graphqlSchema = testProject.tmpdir.read('schema.graphql')
  const typegen = testProject.tmpdir.read('typegen.ts')

  return {
    runFirstBuild,
    runReflectPrisma,
    runReflectNexus,
    runSecondBuildStdout,
    graphqlSchema,
    typegen,
  }
}

function setupTestNexusPrismaProject(): TestProject {
  const testProject = setupTestProject({
    packageJson: {
      scripts: {
        'reflect:prisma': 'prisma generate',
        'reflect:nexus': 'ts-node --transpile-only main',
        build: 'tsc',
      },
      dependencies: {
        '@prisma/client': '^2.18.0',
        '@types/node': '^14.14.32',
        graphql: '^15.5.0',
        nexus: '^1.0.0',
        prisma: '^2.18.0',
        'ts-node': '^9.1.1',
        typescript: '^4.2.3',
      },
    },
  })

  Execa.commandSync(`yalc publish --no-scripts`)
  testProject.run(`yalc add nexus-prisma`)
  testProject.run(`npm install`)

  return testProject
}

let testProject: TestProject

beforeAll(() => {
  assertBuildPresent()
  testProject = setupTestNexusPrismaProject()
})

it('When bundled custom scalars are used the project type checks and generates expected GraphQL schema', () => {
  setupTestProjectCase({
    testProject,
    main: `
      import { makeSchema, objectType } from 'nexus'
      import { M1 } from 'nexus-prisma'
      import * as customScalars from 'nexus-prisma/scalars'
      import * as Path from 'path'
      
      const SomeObject = objectType({
        name: 'SomeObject',
        definition(t) {
          t.json('JsonManually')
          t.dateTime('DateTimeManually')
          t.field(M1.someJsonField.name, M1.someJsonField)
          t.field(M1.someDateTimeField.name, M1.someDateTimeField)
        },
      })
      
      makeSchema({
        types: [customScalars, SomeObject],
        shouldGenerateArtifacts: true,
        shouldExitAfterGenerateArtifacts: true,
        outputs: {
          schema: true,
          typegen: Path.join(__dirname, 'typegen.ts'),
        },
      })

      // wait for output generation
      setTimeout(() => {}, 1000)
    `,
    prismaSchema: createPrismaSchema(`
      model M1 {
        id                String   @id
        someJsonField     Json
        someDateTimeField DateTime
      }
    `),
  })

  // uncomment this to see dir (helpful to go there yourself and manually debug)
  // console.log(testProject.tmpdir.cwd())

  const results = runTestProject(testProject)

  // uncomment this to see the raw results (helpful for debugging)
  // dump(results)

  expect(results.runFirstBuild.exitCode).toBe(2)

  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2305: Module '"nexus-prisma"' has no exported member 'M1'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2339: Property 'json' does not exist on type 'ObjectDefinitionBlock<"SomeObject">'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2339: Property 'dateTime' does not exist on type 'ObjectDefinitionBlock<"SomeObject">'.*/
  )

  expect(results.runReflectPrisma.exitCode).toBe(0)

  expect(stripAnsi(results.runReflectPrisma.stdout)).toMatch(/.*Generated Nexus Prisma.*/)

  expect(results.runReflectNexus.exitCode).toBe(0)

  expect(stripAnsi(results.runReflectNexus.stdout)).toMatch(/.*Generated Artifacts.*/)

  expect(results.runSecondBuildStdout.exitCode).toBe(0)

  expect(results.graphqlSchema).toMatchSnapshot()

  expect(results.typegen).toMatchSnapshot()
})
