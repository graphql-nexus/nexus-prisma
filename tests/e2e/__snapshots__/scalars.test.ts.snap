// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When bundled custom scalars are used the project type checks and generates expected GraphQL schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


\\"\\"\\"
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
\\"\\"\\"
scalar DateTime

\\"\\"\\"
The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
\\"\\"\\"
scalar Json

type Query {
  ok: Boolean!
}

type SomeObject {
  DateTimeManually: DateTime
  JsonManually: Json
  someDateTimeField: DateTime!
  someJsonField: Json!
}
"
`;

exports[`When bundled custom scalars are used the project type checks and generates expected GraphQL schema 2`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */


import { core } from \\"nexus\\"
declare global {
  interface NexusGenCustomInputMethods<TypeName extends string> {
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    dateTime<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"DateTime\\";
    /**
     * The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
     */
    json<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"Json\\";
  }
}
declare global {
  interface NexusGenCustomOutputMethods<TypeName extends string> {
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    dateTime<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"DateTime\\";
    /**
     * The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
     */
    json<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"Json\\";
  }
}


declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  DateTime: any
  Json: any
}

export interface NexusGenObjects {
  Query: {};
  SomeObject: { // root type
    DateTimeManually?: NexusGenScalars['DateTime'] | null; // DateTime
    JsonManually?: NexusGenScalars['Json'] | null; // Json
    someDateTimeField: NexusGenScalars['DateTime']; // DateTime!
    someJsonField: NexusGenScalars['Json']; // Json!
  }
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars

export interface NexusGenFieldTypes {
  Query: { // field return type
    ok: boolean; // Boolean!
  }
  SomeObject: { // field return type
    DateTimeManually: NexusGenScalars['DateTime'] | null; // DateTime
    JsonManually: NexusGenScalars['Json'] | null; // Json
    someDateTimeField: NexusGenScalars['DateTime']; // DateTime!
    someJsonField: NexusGenScalars['Json']; // Json!
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    ok: 'Boolean'
  }
  SomeObject: { // field return type name
    DateTimeManually: 'DateTime'
    JsonManually: 'Json'
    someDateTimeField: 'DateTime'
    someJsonField: 'Json'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    isTypeOf: false
    resolveType: true
    __typename: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
  }
}"
`;
